overlord/registrystate/registrymgr.go

```diff
@@ -0,0 +1,388 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package registrystate
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"gopkg.in/tomb.v2"
+
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/hookstate"
+	"github.com/snapcore/snapd/overlord/ifacestate/ifacerepo"
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+func init() {
+	hookstate.ViewChangedHandlerGenerator = func(context *hookstate.Context) hookstate.Handler {
+		return &viewChangedHandler{ctx: context}
+	}
+	hookstate.SaveViewHandlerGenerator = func(context *hookstate.Context) hookstate.Handler {
+		return &saveViewHandler{ctx: context}
+	}
+	hookstate.ChangeViewHandlerGenerator = func(context *hookstate.Context) hookstate.Handler {
+		return &changeViewHandler{ctx: context}
+	}
+}
+
+func Manager(st *state.State, runner *state.TaskRunner) *RegistryManager {
+	m := &RegistryManager{registryChans: make(map[string]chan struct{})}
+	registryMgr = m
+
+	// no undo since if we commit there's no rolling back
+	runner.AddHandler("commit-transaction", m.doCommitTransaction, nil)
+	// only activated on undo, to clear the ongoing transaction from state and
+	// unblock others who may be waiting for it
+	runner.AddHandler("clear-tx-on-error", m.noop, m.clearOngoingTransaction)
+	runner.AddHandler("clear-tx", m.clearOngoingTransaction, nil)
+
+	return m
+}
+
+var registryMgr *RegistryManager
+
+type RegistryManager struct {
+	// should only be written/read holding the state lock
+	registryChans map[string]chan struct{}
```

Comment:
@pedronis the new waiting mechanism is mostly implemented here and in the helpers below (for when you review)

Reply:
looking at it quickly seemed reasonable